include/range/v3/view/cartesian_product.hpp

```diff
@@ -0,0 +1,377 @@
+/// \file
+// Range v3 library
+//
+//  Copyright Eric Niebler 2013-2014.
+//  Copyright Casey Carter 2017.
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+
+#ifndef RANGES_V3_VIEW_CARTESIAN_PRODUCT_HPP
+#define RANGES_V3_VIEW_CARTESIAN_PRODUCT_HPP
+
+#include <range/v3/begin_end.hpp>
+#include <range/v3/range_access.hpp>
+#include <range/v3/range_concepts.hpp>
+#include <range/v3/range_fwd.hpp>
+#include <range/v3/range_traits.hpp>
+#include <range/v3/size.hpp>
+#include <range/v3/view_facade.hpp>
+#include <range/v3/utility/concepts.hpp>
+#include <range/v3/utility/static_const.hpp>
+#include <range/v3/utility/tuple_algorithm.hpp>
+#include <range/v3/view/all.hpp>
+
+namespace ranges
+{
+    inline namespace v3
+    {
+        namespace detail
+        {
+            template<typename State, typename Value>
+            using product_cardinality =
+                std::integral_constant<cardinality,
+                    State::value == 0 || Value::value == 0 ?
+                        static_cast<cardinality>(0) :
+                        State::value == unknown || Value::value == unknown ?
+                            unknown :
+                            State::value == infinite || Value::value == infinite ?
+                                infinite :
+                                State::value == finite || Value::value == finite ?
+                                    finite :
+                                    static_cast<cardinality>(State::value * Value::value)>;
+
+            struct cartesian_size_fn
+            {
+                template<typename Rng, CONCEPT_REQUIRES_(SizedRange<Rng>())>
+                auto operator()(std::size_t s, Rng && rng)
+                RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
+                (
+                    s * static_cast<std::size_t>(ranges::size(rng))
+                )
+            };
+
+            template<typename View,
+                CONCEPT_REQUIRES_(BoundedRange<View>())>
+            static iterator_t<View> last_iter(View& v)
+            {
+                return ranges::end(v);
+            }
+            template<typename View,
+                CONCEPT_REQUIRES_(!BoundedRange<View>() &&
+                    RandomAccessRange<View>() && SizedRange<View>())>
+            static iterator_t<View> last_iter(View& v)
+            {
+                return ranges::begin(v) + ranges::size(v);
+            }
```

Comment:
Yes; the only difference is that `last_iter` avoids calling `begin` on `BoundedRange`s, and is always O(1).

`last_iter` is only used in `cursor::prev_`, which is only called by `cursor::prev`, which requires `CanBidi`, which basically implies that `ranges::next(begin(v), end(v))` can use one of the O(1) implementations of `ranges::advance`. So I can dump `last_iter` and replace it with a call to `next`.